trin-core/src/portalnet/overlay_service.rs

```diff
@@ -1327,35 +1334,37 @@ where
                 return;
             }
         };
-        let should_store = self.storage.read().should_store(&content_key);
-        match should_store {
-            Ok(val) => {
-                if val {
-                    let validator = Arc::clone(&self.validator);
-                    let storage = Arc::clone(&self.storage);
-                    // Spawn task that validates content before storing.
-                    // Allows for non-blocking requests to this/other overlay services.
-                    tokio::spawn(async move {
-                        if let Err(err) = validator
-                            .validate_content(&content_key, &content.to_vec())
-                            .await
-                        {
-                            warn!("Unable to validate received content: {err:?}");
-                            return;
-                        };
 
-                        if let Err(err) = storage.write().store(&content_key, &content.into()) {
-                            error!("Content received, but not stored: {err}")
-                        }
-                    });
-                } else {
-                    debug!(
-                        "Content received, but not stored: Content is already stored or its distance falls outside current radius."
-                    )
-                }
+        if !self.store.read().is_key_within_radius(&content_key) {
+            debug!("Content received, but not stored: distance falls outside current radius.");
+            return;
+        }
+        let content_lookup = self.store.read().get(&content_key);
+        match content_lookup {
+            Ok(None) => {
+                let validator = Arc::clone(&self.validator);
+                let store = Arc::clone(&self.store);
+                // Spawn task that validates content before storing.
+                // Allows for non-blocking requests to this/other overlay services.
+                tokio::spawn(async move {
+                    if let Err(err) = validator
+                        .validate_content(&content_key, &content.to_vec())
+                        .await
+                    {
+                        warn!("Unable to validate received content: {err:?}");
+                        return;
+                    };
+
+                    if let Err(err) = store.write().put(content_key, content.to_vec()) {
+                        error!("Content received, but not stored: {err}")
+                    }
+                });
             }
-            Err(_) => {
-                error!("Content received, but not stored: Error communicating with db.");
+            Ok(Some(_)) => {
+                debug!("Content received, but not stored: content already stored.");
```

Comment:
Similar here, this error handling seems more appropriate to be handled at the storage layer.